File: src/main/java/co/fastestpath/api/schedule/AgencyId.java
package co.fastestpath.api.schedule;

public class AgencyId {

  private final String value;

  public AgencyId(String value) {
    this.value = value;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    AgencyId agencyId = (AgencyId) o;

    return value != null ? value.equals(agencyId.value) : agencyId.value == null;
  }

  @Override
  public int hashCode() {
    return value != null ? value.hashCode() : 0;
  }
}

File: core/src/main/kotlin/co/fastestpath/api/bootstrap/schedule/AgencyId.kt
package co.fastestpath.api.bootstrap.schedule

data class AgencyId private constructor(val value: String) {
  companion object {
    private val cache: MutableMap<String, AgencyId> = HashMap()

    fun of(value: String): AgencyId {
      var agencyId = cache[value]
      if (agencyId != null) {
        return agencyId
      }
      agencyId = AgencyId(value)
      cache[value] = agencyId
      return agencyId
    }
  }
}
